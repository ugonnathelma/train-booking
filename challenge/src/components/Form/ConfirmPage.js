import React from "react";

import {
  Container,
  Space,
  Button,
  Label,
  FormWrap,
  Flex,
  Detail
} from "./style";

const ConfirmPage = ({
  selectedClass,
  selectedSeat,
  firstName,
  lastName,
  date,
  time,
  price,
  currentTimeLeft,
  stopTimer,
  bookingSuccess,
  cancelBooking
}) => {
  return (
    <Container column>
      <FormWrap>
        <Label>Time left to pay:</Label> <Detail time>{currentTimeLeft}</Detail>
        <Space vertical="0.8em" />
        <Label>Class:</Label> <Detail small>{selectedClass}</Detail>
        <Space vertical="0.8em" />
        <Label>Seat:</Label>
        <Detail small>{selectedSeat}</Detail>
        <Space vertical="0.8em" />
        <Label>Passenger Name:</Label>{" "}
        <Detail small>{`${firstName} ${lastName}`}</Detail>
        <Space vertical="0.8em" />
        <Label>Travel Date:</Label> <Detail small>{date}</Detail>
        <Space vertical="0.8em" />
        <Label>Travel Time:</Label> <Detail small>{time}</Detail>
        <Space vertical="0.8em" />
        <Label>Price:</Label> <Detail small>{price}</Detail>
        <Space vertical="1.5em" />
        <Flex>
          <Button
            width="10em"
            onClick={() => {
              stopTimer();
              bookingSuccess(true);
            }}
          >
            Pay
          </Button>
          <Space horizontal="1.5em" />
          <Button width="10em" onClick={() => cancelBooking()}>
            Cancel
          </Button>
        </Flex>
      </FormWrap>
    </Container>
  );
};

export default ConfirmPage;
