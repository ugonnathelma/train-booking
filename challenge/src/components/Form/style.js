import styled from "styled-components";
import TrainPic from "../../images/train.jpg";

const theme = {
  primaryColor: "#c44569",
  tintColor: "#cf6a87",
  selectColor: "#2980b9",
  grey: "gainsboro"
};

export const RadioButton = styled.input`
  cursor: pointer;
`;

export const Container = styled.section`
  display: flex;
  ${({ column }) => column && `flex-direction: column`};
  width: 100%;
  height: 100vh;

  * {
    box-sizing: border-box;
  }
`;

export const LeftPanel = styled.section`
  flex: 1;
  padding: 2em;
`;

export const RightPanel = styled.section`
  flex: 1;
  background: url("${TrainPic}");
  background-size:cover;
  background-repeat:no-repeat;
`;

export const FormWrap = styled.section`
  flex: 1;
  display: flex;
  flex-direction: column;

  h2 {
    margin: 0;
  }
`;

export const Flex = styled.section`
  display: flex;
  align-items: center;
  ${({ column }) => column && `flex-direction: column`};

  ${({ inputs }) =>
    inputs &&
    `
  @media (max-width: 500px) {
    flex-direction: column !important;
    align-items: flex-start;

    ${FlexChild} {
      flex:1
    }
  }`};
`;

export const FlexChild = styled.section`
  flex: ${({ flex }) => (flex ? flex : 1)};
  display: flex;
  flex-direction: column;
  width: 100%;
`;

export const Space = styled.div`
  ${({ horizontal }) => horizontal && `width: ${horizontal}`};
  ${({ vertical }) => vertical && `height: ${vertical}`}
`;

export const Seat = styled.div`
  background: ${({ available }) =>
    available ? theme.selectColor : theme.grey};
  width: 2em;
  height: 2em;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 1em;
  cursor: pointer;

  ${({ active, small }) =>
    active &&
    !small &&
    `box-shadow:inset 0px 0px 0px 3px ${theme.primaryColor}`};
  ${({ small }) =>
    small &&
    `width: 1em;
  height: 1em;`};

  ${({ available }) => available && "color:white"};
`;

export const SeatSection = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Button = styled.button`
  background: ${({ disabled }) => (disabled ? theme.grey : theme.primaryColor)};
  color: white;
  padding: 1em;
  outline: none;
  border: none;
  max-width: 20em;
  cursor: pointer;
  ${({ width }) => width && `width:${width}`};

  transition: background 0.3s;

  &:hover {
    background: ${({ disabled }) => (disabled ? theme.grey : theme.tintColor)};
  }
`;

export const TextField = styled.input`
  padding: 0.7em;
  border: 1px solid gainsboro;
  outline: none;
  font-size: 14px;
  width: 100%;
`;

export const Label = styled.label`
  font-size: 16px;
  font-weight: bold;
`;

export const Hr = styled.div`
  border-bottom: 1px solid ${theme.grey};
`;

export const Detail = styled.label`
  font-size: ${({ small }) => (small ? "18px" : "45px")};
  color: ${theme.selectColor};

  ${({ time }) =>
    time &&
    `
  
  `};
`;
