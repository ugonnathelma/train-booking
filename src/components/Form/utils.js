export const firstClassSeats = {
  a: [
    { number: 1, available: true },
    { number: 2, available: true },
    { number: 3, available: false },
    { number: 4, available: false },
    { number: 5, available: true }
  ],
  b: [
    { number: 6, available: true },
    { number: 7, available: false },
    { number: 8, available: true },
    { number: 9, available: true },
    { number: 10, available: false }
  ]
};

export const secondClassSeats = {
  c: [
    { number: 11, available: true },
    { number: 12, available: true },
    { number: 13, available: true },
    { number: 14, available: true },
    { number: 15, available: true }
  ],
  d: [
    { number: 16, available: true },
    { number: 17, available: false },
    { number: 18, available: true },
    { number: 19, available: true },
    { number: 20, available: true }
  ]
};
