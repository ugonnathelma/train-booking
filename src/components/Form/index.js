import React, { useState } from "react";
import moment from "moment";
import ConfirmPage from "./ConfirmPage";
import {
  RadioButton,
  Container,
  Space,
  FormWrap,
  Flex,
  SeatSection,
  Seat,
  Button,
  TextField,
  Label,
  FlexChild,
  LeftPanel,
  RightPanel,
  Hr
} from "./style";
import { firstClassSeats, secondClassSeats } from "./utils";

let timerRef;

const startTimer = (setTimer, cancelBooking, minutes, seconds) => {
  timerRef = setInterval(() => {
    if (seconds === 0) {
      minutes = minutes - 1;
      seconds = 59;
    } else {
      seconds = seconds - 1;
    }
    if (seconds === 0 && minutes === 0) {
      stopTimer();
      cancelBooking();
    }
    setTimer({ minutes, seconds });
  }, 1000);
};

const stopTimer = () => {
  clearInterval(timerRef);
};

const Form = () => {
  const [selectedClass, selectClass] = useState(1);
  const [selectedSeat, selectSeat] = useState(null);
  const [showConfirmPage, setShowConfirm] = useState(false);
  const [success, setSuccess] = useState(false);
  const [cancelled, setCancel] = useState(false);
  const [timer, setTimer] = useState({
    minutes: 3,
    seconds: 0
  });

  const [passengerDetails, setPassengerDetails] = useState({
    firstName: "",
    lastName: "",
    date: moment().format("YYYY-MM-DD"),
    time: null
  });

  const isFirstClass = selectedClass === 1;

  const seats = isFirstClass ? firstClassSeats : secondClassSeats;

  const { firstName, lastName, date, time } = passengerDetails;

  const checkInButtonDisabled = !(
    selectedClass &&
    selectedSeat &&
    firstName.length &&
    lastName.length &&
    date &&
    time
  );

  const setInput = (value, key) =>
    setPassengerDetails({ ...passengerDetails, [key]: value });

  const cancelBooking = () => {
    setCancel(true);
    setTimer({
      minutes: 3,
      seconds: 0
    });
  };
  const reset = () => {
    setCancel(false);
    setSuccess(false);
    setShowConfirm(false);
    selectSeat(null);
    selectClass(1);
    setTimer({
      minutes: 3,
      seconds: 0
    });

    setPassengerDetails({
      firstName: "",
      lastName: "",
      date: moment().format("YYYY-MM-DD"),
      time: null
    });
  };

  return (
    <Container>
      <LeftPanel slideRight={showConfirmPage}>
        <h2>Train trip booking</h2>
        <Hr></Hr>
        <Space vertical="2em" />
        {!showConfirmPage && (
          <FormWrap>
            <Flex>
              <label>available:</label> <Seat available={true} small></Seat>{" "}
              <label>taken:</label>
              <Seat small></Seat>
            </Flex>
            <Space vertical="2em" />
            <Label>Please choose a class</Label>
            <Space vertical="1em" />
            <Flex>
              <RadioButton
                readOnly
                type="radio"
                name="class"
                value="first"
                checked={isFirstClass}
                onClick={() => selectClass(1)}
              />{" "}
              First Class
              <Space horizontal="2em" />
              <RadioButton
                readOnly
                type="radio"
                name="class"
                value="second"
                checked={!isFirstClass}
                onClick={() => selectClass(2)}
              />{" "}
              Second Class
            </Flex>
            {Object.keys(seats).map(carriage => (
              <div key={carriage}>
                <Space vertical="1em" />
                <Label>Carriage:</Label> {carriage.toUpperCase()}
                <SeatSection>
                  {seats[carriage].map(seat => (
                    <Seat
                      key={seat.number}
                      available={seat.available}
                      active={
                        selectedSeat === carriage.toUpperCase() + seat.number
                      }
                      onClick={() =>
                        seat.available &&
                        selectSeat(carriage.toUpperCase() + seat.number)
                      }
                    >
                      {seat.number}
                    </Seat>
                  ))}
                </SeatSection>
              </div>
            ))}
            <Flex inputs>
              <FlexChild>
                <Label htmlFor="firstName">First Name</Label>{" "}
                <TextField
                  id="firstName"
                  value={firstName}
                  onChange={event => setInput(event.target.value, "firstName")}
                />
              </FlexChild>
              <Space vertical="1em" horizontal="3em" />
              <FlexChild>
                <Label htmlFor="lastName">Last Name</Label>
                <TextField
                  id="lastName"
                  value={lastName}
                  onChange={event => setInput(event.target.value, "lastName")}
                />
              </FlexChild>
            </Flex>
            <Space vertical="1em" />
            <Flex inputs>
              <FlexChild>
                <Label htmlFor="travelDate">Travel Date</Label>
                <TextField
                  id="travelDate"
                  value={date}
                  type="date"
                  onChange={event => setInput(event.target.value, "date")}
                />
              </FlexChild>
              <Space vertical="1em" horizontal="3em" />
              <FlexChild>
                <Label htmlFor="travelTime">Travel Time</Label>
                <TextField
                  id="travelTime"
                  value={time}
                  type="time"
                  onChange={event => setInput(event.target.value, "time")}
                />
              </FlexChild>
            </Flex>
            <Space vertical="1em" />
            <Button
              disabled={checkInButtonDisabled}
              onClick={() => {
                startTimer(
                  setTimer,
                  cancelBooking,
                  timer.minutes,
                  timer.seconds
                );
                !checkInButtonDisabled && setShowConfirm(true);
              }}
            >
              Check in
            </Button>
          </FormWrap>
        )}{" "}
        <div>
          {showConfirmPage && !cancelled && !success ? (
            <ConfirmPage
              selectedClass={selectedClass === 1 ? "First" : "Second"}
              selectedSeat={selectedSeat}
              {...passengerDetails}
              currentTimeLeft={`${timer.minutes}:${(timer.seconds < 10
                ? "0"
                : "") + timer.seconds}`}
              stopTimer={stopTimer}
              bookingSuccess={setSuccess}
              cancelBooking={cancelBooking}
              price={selectedClass === 1 ? "€150" : "€65"}
            />
          ) : (
            success &&
            !cancelled && (
              <div>
                <Label success>Reservation is successful</Label>
                <Space vertical="1em" />
                <Button onClick={() => cancelBooking()}>Cancel</Button>
              </div>
            )
          )}
          {cancelled && (
            <div>
              <Label>Reservation is cancelled</Label>
              <Space vertical="1em" />
              <Button width="7em" onClick={() => reset()}>
                Back
              </Button>
            </div>
          )}
        </div>
      </LeftPanel>
      <RightPanel slideLeft={showConfirmPage}></RightPanel>
    </Container>
  );
};

export default Form;
