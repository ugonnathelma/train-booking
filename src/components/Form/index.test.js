import React from "react";
import Form from "./index";
import { render } from "@testing-library/react";

describe("Form", () => {
  it("should render matching the existing snapshot", () => {
    const { asFragment } = render(<Form />);
    expect(asFragment()).toMatchSnapshot();
  });
});
