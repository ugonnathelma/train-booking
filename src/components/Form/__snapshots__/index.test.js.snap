// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render matching the existing snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-bwzfXH iAPMsk"
  >
    <section
      class="sc-htpNat iyhOGH"
    >
      <h2>
        Train trip booking
      </h2>
      <div
        class="sc-VigVT kwmnKF"
      />
      <div
        class="sc-gzVnrw derNAl"
      />
      <section
        class="sc-ifAKCX huQwQB"
      >
        <section
          class="sc-EHOje cIxLkv"
        >
          <label>
            available:
          </label>
           
          <div
            class="sc-htoDjs bKRqCk"
          />
           
          <label>
            taken:
          </label>
          <div
            class="sc-htoDjs kfhMrO"
          />
        </section>
        <div
          class="sc-gzVnrw derNAl"
        />
        <label
          class="sc-gqjmRU fxwXMY"
        >
          Please choose a class
        </label>
        <div
          class="sc-gzVnrw cZOpws"
        />
        <section
          class="sc-EHOje cIxLkv"
        >
          <input
            checked=""
            class="sc-bdVaJa fVCJpx"
            name="class"
            readonly=""
            type="radio"
            value="first"
          />
           First Class
          <div
            class="sc-gzVnrw ceXOul"
          />
          <input
            class="sc-bdVaJa fVCJpx"
            name="class"
            readonly=""
            type="radio"
            value="second"
          />
           Second Class
        </section>
        <div>
          <div
            class="sc-gzVnrw cZOpws"
          />
          <label
            class="sc-gqjmRU fxwXMY"
          >
            Carriage:
          </label>
           A
          <div
            class="sc-dnqmqq eWlhsD"
          >
            <div
              class="sc-htoDjs hPgVdR"
            >
              1
            </div>
            <div
              class="sc-htoDjs hPgVdR"
            >
              2
            </div>
            <div
              class="sc-htoDjs eaKqHr"
            >
              3
            </div>
            <div
              class="sc-htoDjs eaKqHr"
            >
              4
            </div>
            <div
              class="sc-htoDjs hPgVdR"
            >
              5
            </div>
          </div>
        </div>
        <div>
          <div
            class="sc-gzVnrw cZOpws"
          />
          <label
            class="sc-gqjmRU fxwXMY"
          >
            Carriage:
          </label>
           B
          <div
            class="sc-dnqmqq eWlhsD"
          >
            <div
              class="sc-htoDjs hPgVdR"
            >
              6
            </div>
            <div
              class="sc-htoDjs eaKqHr"
            >
              7
            </div>
            <div
              class="sc-htoDjs hPgVdR"
            >
              8
            </div>
            <div
              class="sc-htoDjs hPgVdR"
            >
              9
            </div>
            <div
              class="sc-htoDjs eaKqHr"
            >
              10
            </div>
          </div>
        </div>
        <section
          class="sc-EHOje levPRg"
        >
          <section
            class="sc-bZQynM cJwquT"
          >
            <label
              class="sc-gqjmRU fxwXMY"
              for="firstName"
            >
              First Name
            </label>
             
            <input
              class="sc-gZMcBi btdWif"
              id="firstName"
              value=""
            />
          </section>
          <div
            class="sc-gzVnrw QhzJ"
          />
          <section
            class="sc-bZQynM cJwquT"
          >
            <label
              class="sc-gqjmRU fxwXMY"
              for="lastName"
            >
              Last Name
            </label>
            <input
              class="sc-gZMcBi btdWif"
              id="lastName"
              value=""
            />
          </section>
        </section>
        <div
          class="sc-gzVnrw cZOpws"
        />
        <section
          class="sc-EHOje levPRg"
        >
          <section
            class="sc-bZQynM cJwquT"
          >
            <label
              class="sc-gqjmRU fxwXMY"
              for="travelDate"
            >
              Travel Date
            </label>
            <input
              class="sc-gZMcBi btdWif"
              id="travelDate"
              type="date"
              value="2019-09-07"
            />
          </section>
          <div
            class="sc-gzVnrw QhzJ"
          />
          <section
            class="sc-bZQynM cJwquT"
          >
            <label
              class="sc-gqjmRU fxwXMY"
              for="travelTime"
            >
              Travel Time
            </label>
            <input
              class="sc-gZMcBi btdWif"
              id="travelTime"
              type="time"
              value=""
            />
          </section>
        </section>
        <div
          class="sc-gzVnrw cZOpws"
        />
        <button
          class="sc-iwsKbI kOYkMQ"
          disabled=""
        >
          Check in
        </button>
      </section>
       
      <div />
    </section>
    <section
      class="sc-bxivhb csVKtc"
    />
  </section>
</DocumentFragment>
`;
